src/core/tsi/alts_transport_security.cc

```diff
@@ -63,3 +64,15 @@ void grpc_tsi_alts_shutdown() {
   gpr_cv_destroy(&g_alts_resource.cv);
   gpr_mu_destroy(&g_alts_resource.mu);
 }
+
+void grpc_tsi_alts_shared_resource_ref() { gpr_ref(&g_alts_resource.refcount); }
```

Comment:
What happens if the application creates a channel and then destroys it (thus sending the refcount to 0) and then creates another channel?  We need some way to recreate the shared resources when the second channel is created.

Reply:
Initialization is lazily at https://github.com/grpc/grpc/blob/master/src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc#L333

Reply:
Let me say this a different way: In terms of the API here, it seems odd to me that `grpc_tsi_alts_shared_resource_ref()` and `alts_get_shared_resource()` are different methods.  Shouldn't the API be that there is a single method that is responsible for returning a pointer along with a new ref, which the caller holds until they explicitly unref it?  Otherwise, the API allows weird things like taking a new ref when the previous refcount was 0 and no object actually exists.

Reply:
I agree it will be better to have an API that returns a pointer along with the ref. Since in the current ALTS impl, the shared resource is needed at the TSI layer, we need to call `grpc_tsi_alts_shared_resource_ref()` at the credential layer and plumb it all the way down to TSI. I will update the code.